lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -598,6 +600,39 @@ newDBLayer trace defaultFieldValues mDatabaseFile timeInterpreter = do
             migrateAll
             trace
             mDatabaseFile
+
+    cache <- newIORef Nothing
+
+    let selectLatestCheckpoint
+            :: W.WalletId
+            -> SqlPersistT IO (Maybe (W.Wallet s))
+        selectLatestCheckpoint wid =
+            liftIO (readIORef cache) >>= maybe fromDatabase (pure . Just)
```

Comment:
The cache doesn't live in `atomically`, in contrast to the rest of the DB layer. Couldn't this lead to race-conditions?

Reply:
There can only be a single execution thread in `atomically` (constructed with an MVar lock), hence the use of a mere IORef here.

Reply:
I was imagining something more like:

```
----------------------------------------------------------------------> time
1: putCheckpoint cp1 
2:           selectLatest >>= f >>= putCheckpoint
                                                     selectLatest >>= g >>= putCheckpoint
```

1. Updates cache, but operation is cancelled before the DB is updated. Cache and DB diverges.
2. DB cp will be overwritten by the cached checkpoint the next time it is modified

But I guess the worst thing that can happen is that a somehow-failing `putCheckpoint` will end up succeeding 🤔

Reply:
> Updates cache, but operation is cancelled before the DB is updated. Cache and DB diverges.

Cache is updated _after_ the put operation has succeeded though. If it doesn't succeed, the cache isn't.

Reply:
Ok, Matthias pointed out that "We create an illusion of transaction by having a top-level MVar lock.", so there shouldn't be Sqlite transactions being run and then somehow cancelled.

Reply:
I also added some notes in the code about this in place where the cache is created.